log/codecomet_writer.go

```diff
@@ -0,0 +1,471 @@
+		buf.WriteByte(' ')
+		buf.WriteByte(' ')
+	}
+
+	// Move the "error" field to the front
+	ei := sort.Search(len(fields), func(i int) bool { return fields[i] >= zerolog.ErrorFieldName })
+	if ei < len(fields) && fields[ei] == zerolog.ErrorFieldName {
+		fields[ei] = ""
+		fields = append([]string{zerolog.ErrorFieldName}, fields...)
+		xfields := make([]string, 0, len(fields))
+		for _, field := range fields {
+			if field == "" { // Skip empty fields
+				continue
+			}
+			xfields = append(xfields, field)
+		}
+		fields = xfields
+	}
+
+	for i, field := range fields {
+		var fn Formatter
+		var fv Formatter
+
+		if field == zerolog.ErrorFieldName {
+			if w.FormatErrFieldName == nil {
+				fn = consoleDefaultFormatErrFieldName(w.NoColor)
+			} else {
+				fn = w.FormatErrFieldName
+			}
+
+			if w.FormatErrFieldValue == nil {
+				fv = consoleDefaultFormatErrFieldValue(w.NoColor)
+			} else {
+				fv = w.FormatErrFieldValue
+			}
+		} else {
+			if w.FormatFieldName == nil {
+				fn = consoleDefaultFormatFieldName(w.NoColor)
+			} else {
+				fn = w.FormatFieldName
+			}
+
+			if w.FormatFieldValue == nil {
+				fv = consoleDefaultFormatFieldValue
+			} else {
+				fv = w.FormatFieldValue
+			}
+		}
+
+		buf.WriteString(fn(field))
+
+		switch fValue := evt[field].(type) {
+		case string:
+			if needsQuote(fValue) {
+				buf.WriteString(fv(strconv.Quote(fValue)))
+			} else {
+				buf.WriteString(fv(fValue))
+			}
+		case json.Number:
+			buf.WriteString(fv(fValue))
+		default:
+			b, err := zerolog.InterfaceMarshalFunc(fValue)
+			if err != nil {
+				fmt.Fprintf(buf, colorize("[error: %v]", colorRed, w.NoColor), err)
+			} else {
+				fmt.Fprint(buf, fv(b))
+			}
+		}
+
+		if i < len(fields)-1 { // Skip space for last field
+			buf.WriteByte(' ')
+		}
+	}
+}
+
+// writePart appends a formatted part to buf.
+func (w CodecometWriter) writePart(buf *bytes.Buffer, evt map[string]interface{}, p string) {
+	var f Formatter
+
+	if w.PartsExclude != nil && len(w.PartsExclude) > 0 {
+		for _, exclude := range w.PartsExclude {
+			if exclude == p {
+				return
+			}
+		}
+	}
+
+	switch p {
+	case zerolog.LevelFieldName:
+		if w.FormatLevel == nil {
+			f = consoleDefaultFormatLevel(w.NoColor)
+		} else {
+			f = w.FormatLevel
+		}
+	case zerolog.TimestampFieldName:
+		if w.FormatTimestamp == nil {
+			f = consoleDefaultFormatTimestamp(w.TimeFormat, w.NoColor)
+		} else {
+			f = w.FormatTimestamp
+		}
+	case zerolog.MessageFieldName:
+		if w.FormatMessage == nil {
+			f = consoleDefaultFormatMessage
+		} else {
+			f = w.FormatMessage
+		}
+	case ContextFieldName:
+		if w.FormatContext == nil {
+			f = consoleDefaultFormatContext
+		} else {
+			f = w.FormatContext
+		}
+	case ModeFieldName:
+		if w.FormatMode == nil {
+			f = consoleDefaultFormatMode
+		} else {
+			f = w.FormatMode
+		}
+	// case zerolog.CallerFieldName:
+	// 	if w.FormatCaller == nil {
+	// 		f = consoleDefaultFormatCaller(w.NoColor)
+	// 	} else {
+	// 		f = w.FormatCaller
+	// 	}
```

Comment:
annoyingly, i spent time yesterday restoring the caller functionality so that if anyone wanted to use it, they could... but somehow, my setup here lost that change [still grappling with it]. alas. I can restore this again later. not going to hold this up now for it. :(